packages/xo-web/src/common/modal.js

```diff
@@ -259,57 +260,103 @@ export const confirm = ({ body, icon = 'alarm', title, strongConfirm }) =>
 
 // -----------------------------------------------------------------------------
 
-const preventDefault = event => event.preventDefault()
-
-class FormModal extends BaseComponent {
-  state = {
-    value: this.props.defaultValue,
-  }
-
-  get value () {
-    return this.state.value
-  }
-
-  render () {
-    const { body, formId } = this.props
-    return (
-      <form id={formId} onSubmit={preventDefault}>
-        {cloneElement(body, {
-          value: this.state.value,
-          onChange: this.linkState('value'),
-        })}
-      </form>
-    )
-  }
-}
-
-export const form = ({ body, defaultValue, icon, title, size }) => {
-  const formId = generateRandomId()
-  const buttons = [
-    {
-      btnStyle: 'primary',
-      label: _('formOk'),
-      form: formId,
-    },
-  ]
-  return new Promise((resolve, reject) => {
-    modal(
-      <GenericModal
-        buttons={buttons}
-        icon={icon}
-        reject={reject}
-        resolve={resolve}
-        title={title}
-      >
-        <FormModal body={body} defaultValue={defaultValue} formId={formId} />
-      </GenericModal>,
+let formModalState
+export const form = props =>
+  new Promise((resolve, reject) => {
+    formModalState.props = {
       reject,
-      {
-        bsSize: size,
-      }
-    )
+      resolve,
+      showModal: true,
+      ...props,
+    }
+    disableShortcuts()
   })
-}
+
+const getInitialState = () => ({
+  id: generateRandomId(),
+  localValue: undefined,
+  props: undefined,
+})
+export const FormModal = [
+  provideState({
+    initialState: getInitialState,
+    effects: {
+      initialize () {
+        formModalState === undefined && (formModalState = this.state)
```

Comment:
1. use explicit `if` in this case
2. IMO it's better to throw an error if `formModalState` is already defined, otherwise the component might silently incorrectly